File: examples/ruby/solid_jobs_client.rb
require 'net/http'
require 'json'
require 'uri'

class SolidJobsClient
  CAMPAIGN_RE = /\A[a-z0-9-]{1,64}\z/.freeze

  def initialize(base_url: 'https://solid.jobs', api_version: '1.0', max_retries: 3)
    @base_url = base_url
    @api_version = api_version
    @max_retries = max_retries
  end

  def get_offers(division, campaign, **query)
    raise ArgumentError, 'Invalid campaign format' unless CAMPAIGN_RE.match?(campaign)

    url = build_url(division, campaign, query)
    body = request_with_retry(url)
    JSON.parse(body)
  end

  def get_all_offers(division, campaign, **query)
    page_index = query.fetch(:page_index, 0)
    all = []

    loop do
      page = get_offers(division, campaign, **query.merge(page_index: page_index))
      all.concat(page['jobs'])
      page_index += 1
      break if page['jobs'].empty? || page_index >= page['totalPages']
    end

    all
  end

  def get_market_statistics(scope_kind, scope_key, campaign, fields: nil)
    raise ArgumentError, 'Invalid campaign format' unless CAMPAIGN_RE.match?(campaign)

    params = [['campaign', campaign]]
    params << ['fields', Array(fields).join(',')] if fields && !fields.empty?

    query_string = URI.encode_www_form(params)
    url = "#{@base_url}/public-api/market-statistics/#{scope_kind}/#{scope_key}?#{query_string}"
    body = request_with_retry(url)
    JSON.parse(body)
  end

  private

  def request_with_retry(url)
    uri = URI(url)

    (0..@max_retries).each do |attempt|
      response = Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) do |http|
        req = Net::HTTP::Get.new(uri)
        req['X-Api-Version'] = @api_version
        http.request(req)
      end

      unless response.code == '429' && attempt < @max_retries
        raise "HTTP #{response.code}: #{response.body}" unless response.code == '200'
        return response.body
      end

      delay = response['Retry-After']&.to_i || (2**attempt)
      warn "Rate limited (429). Retrying in #{delay}s..."
      sleep(delay)
    end
  end

  def build_url(division, campaign, query)
    params = [['campaign', campaign]]

    params << ['pageIndex', query[:page_index]] if (query[:page_index] || 0) > 0
    params << ['pageSize', query[:page_size]] if query[:page_size]

    if query[:sort_active]
      params << ['sortActive', query[:sort_active]]
      params << ['sortDirection', query[:sort_direction] || 'asc']
    end

    params << ['search.cities', query[:cities].join(',')] if query[:cities]&.any?

    { categories: 'search.categories', sub_categories: 'search.subCategories',
      experiences: 'search.experiences', search_terms: 'search.searchTerm' }.each do |key, param|
      (query[key] || []).each { |v| params << [param, v] }
    end

    params << ['search.minimumSalary', query[:minimum_salary]] if query[:minimum_salary]

    query_string = URI.encode_www_form(params)
    "#{@base_url}/public-api/offers/#{division}?#{query_string}"
  end
end

File: examples/php/SolidJobsClient.php
<?php

class SolidJobsClient
{
    private string $baseUrl;
    private string $apiVersion;
    private int $maxRetries;

    public function __construct(
        string $baseUrl = 'https://solid.jobs',
        string $apiVersion = '1.0',
        int $maxRetries = 3
    ) {
        $this->baseUrl = $baseUrl;
        $this->apiVersion = $apiVersion;
        $this->maxRetries = $maxRetries;
    }

    public function getOffers(string $division, string $campaign, array $query = []): array
    {
        if (!preg_match('/^[a-z0-9-]{1,64}$/', $campaign)) {
            throw new InvalidArgumentException('Campaign must contain only lowercase letters, digits and hyphens (max 64 chars).');
        }

        $url = $this->buildUrl($division, $campaign, $query);
        $body = $this->requestWithRetry($url);

        return json_decode($body, true);
    }

    public function getAllOffers(string $division, string $campaign, array $query = []): array
    {
        $all = [];
        $pageIndex = $query['pageIndex'] ?? 0;

        while (true) {
            $query['pageIndex'] = $pageIndex;
            $page = $this->getOffers($division, $campaign, $query);

            foreach ($page['jobs'] as $job) {
                $all[] = $job;
            }

            $pageIndex++;
            if (empty($page['jobs']) || $pageIndex >= $page['totalPages']) {
                break;
            }
        }

        return $all;
    }

    public function getMarketStatistics(string $scopeKind, string $scopeKey, string $campaign, array $fields = []): array
    {
        if (!preg_match('/^[a-z0-9-]{1,64}$/', $campaign)) {
            throw new InvalidArgumentException('Campaign must contain only lowercase letters, digits and hyphens (max 64 chars).');
        }

        $params = ['campaign' => $campaign];
        if (!empty($fields)) {
            $params['fields'] = implode(',', $fields);
        }

        $url = "{$this->baseUrl}/public-api/market-statistics/"
            . rawurlencode($scopeKind) . '/' . rawurlencode($scopeKey)
            . '?' . http_build_query($params);

        $body = $this->requestWithRetry($url);

        return json_decode($body, true);
    }

    public function getMarketRaport(string $scopeKey, string $campaign): array
    {
        if (!preg_match('/^[a-z0-9-]{1,64}$/', $campaign)) {
            throw new InvalidArgumentException('Campaign must contain only lowercase letters, digits and hyphens (max 64 chars).');
        }

        $params = ['campaign' => $campaign];

        $url = "{$this->baseUrl}/public-api/market-statistics/raport/"
            . rawurlencode($scopeKey)
            . '?' . http_build_query($params);

        $body = $this->requestWithRetry($url);

        return json_decode($body, true);
    }

    private function requestWithRetry(string $url): string
    {
        for ($attempt = 0; ; $attempt++) {
            $ch = curl_init($url);
            curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
            curl_setopt($ch, CURLOPT_TIMEOUT, 30);
            curl_setopt($ch, CURLOPT_HTTPHEADER, ["X-Api-Version: {$this->apiVersion}"]);

            $headers = [];
            curl_setopt($ch, CURLOPT_HEADERFUNCTION, function ($ch, $header) use (&$headers) {
                $parts = explode(':', $header, 2);
                if (count($parts) === 2) {
                    $headers[strtolower(trim($parts[0]))] = trim($parts[1]);
                }
                return strlen($header);
            });

            $body = curl_exec($ch);
            $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
            $curlError = curl_error($ch);
            curl_close($ch);

            if ($curlError) {
                throw new RuntimeException("Network error: $curlError");
            }

            if ($httpCode !== 429 || $attempt >= $this->maxRetries) {
                if ($httpCode !== 200) {
                    throw new RuntimeException("HTTP $httpCode: $body");
                }
                return $body;
            }

            $delay = isset($headers['retry-after']) ? (int) $headers['retry-after'] : pow(2, $attempt);
            fwrite(STDERR, "Rate limited (429). Retrying in {$delay}s...\n");
            sleep($delay);
        }
    }

    private function buildUrl(string $division, string $campaign, array $query): string
    {
        $params = ['campaign' => $campaign];

        if (($query['pageIndex'] ?? 0) > 0) {
            $params['pageIndex'] = $query['pageIndex'];
        }
        if (isset($query['pageSize'])) {
            $params['pageSize'] = $query['pageSize'];
        }
        if (isset($query['sortActive'])) {
            $params['sortActive'] = $query['sortActive'];
            $params['sortDirection'] = $query['sortDirection'] ?? 'asc';
        }
        if (!empty($query['cities'])) {
            $params['search.cities'] = implode(',', $query['cities']);
        }

        $queryString = http_build_query($params);

        foreach (['categories' => 'search.categories', 'subCategories' => 'search.subCategories',
                   'experiences' => 'search.experiences', 'searchTerms' => 'search.searchTerm'] as $key => $param) {
            foreach ($query[$key] ?? [] as $value) {
                $queryString .= '&' . urlencode($param) . '=' . urlencode($value);
            }
        }

        if (isset($query['minimumSalary'])) {
            $queryString .= '&search.minimumSalary=' . (int) $query['minimumSalary'];
        }

        return "{$this->baseUrl}/public-api/offers/{$division}?{$queryString}";
    }
}
